### PR TITLE
tests: fetchTrades: more cases

### PR DESCRIPTION
**Pull request by mkutny**
Some exchanges DO fail on it (so it will hang here until we fix them).

Do you want me to fix them in this PR or to submit a separate PR for every failed exchange?

### CODE REVIEW

**Comment by kroitor** [comment]
> Do you want me to fix them in this PR or to submit a separate PR for every failed exchange?

It would be easier to do fixes, if we make them separate. So, if it isn't very hard for you, the second way would be preferable. Thanks for your help, you're doing great job!

**Comment by mkutny** [comment]
> It would be easier for us to do fixes, if we make them separate.

It's easier for me too.
I've opend PRs - once you merge them this one should pass.

> P.S. Travis is sooo slooooow today...

I guess it's celebrating already!

**Comment by kroitor** [comment]
Also related to #986

**Comment by mkutny** [comment]
As a part of test scenarios on which I'm working now, there are tests for ccxt throwing proper InsufficientFunds and InvalidOrder exceptions.

These kinds of tests require order creation and the safest way to create orders would be using currencies with zero balances (if there are some funds then creating orders exposes users to risks: I personally had issues when an exchange reported an error but still created an order and this just might go unnoticed by a user).

Search for zero-balance currencies on user-side is not straightforward, though.

At first, there might be no such currencies at all - a typical exchange usually provides 3 quote currencies: one fiat (USD/JPY/CNY whatever), BTC and ETH. If a user trades on the exchange he/she quite often has some funds on all of the three.

I also tried to work on a fall-back scenario when 'amount' is set to max limit (but what if there is no max limit?) and the price is set to min limit (but what if an orderbook has ask orders at this price-level too? and what if `price * amount` is less than the `cost` limit?)

In essence, all this discovery logic seems to be overkill for such a trivial test. Thus I thought that ccxt keys would be an ideal option: besides belonging to zero-balance accounts (I hope) they would also allow these kinds of tests to be done on ccxt-side.

I tried some keys a while ago but they had no permissions for order creation. What do you think about enabling that?

**Comment by kroitor** [comment]
@mkutny thanks for posting it!

> In essence, all this discovery logic seems to be overkill for such a trivial test. 

I would agree.

> Thus I thought that ccxt keys would be an ideal choice: besides belonging to zero-balance accounts (I hope) they would also allow these kinds of tests to be done on ccxt-side.

This has several pitfalls: 
1. Someone is constantly screwing my nonces, so having them on Travis does not really solve it completely without encryption... We would have to share some decryption schema among the devs then.
2. Have you considered using sandbox/testnet urls and apis where available? We have some work done in that direction as well (this.urls['api']['test']), but it is still in progress, well, you know ;)
3. Some exchanges have very restrictive KYC/AML rules (like GDAX), and our only option is to test against their sandbox. Some sandboxes, however, drop their temporary states regularly.
4. Multiple users testing simultaneously = they will break each others tests (due to nonce conflicts like those you had with Liqui) + they will interfere with Travis (not solved yet in current code).

Unfortunately, testing with that set can fail if done without automating the signups and key-mgmt via their websites. The biggest pain with testing keys is keeping them up to date... I would be happy to know your thoughts on the above.

**Comment by mkutny** [comment]
What I basically say (beyond rationalie that testing on zero-balance accounts is safer) is that if ccxt keys have trading permissions I will be able to extend ccxt test suite with error handling checks.


> Multiple users testing simultaneously = they will break each others tests

It looks like you assumed that I want to use ccxt keys for user testing, but this is not the case. For user-side testing, I'll make tests with order creation/cancellation logic and that would require some balances in their accounts.

So basically I'd be happy with the following split:
- ccxt accounts/keys for tests that rely on zero balances are going to be a part of ccxt integration tests
- user accounts/keys for tests that rely on tradable balances for user-side testing


> Someone is constantly screwing my nonces

What do you mean by "screwing"? Running examples will impact ccxt nounces, that's expected. But that should not hurt. You mean that someone sets them somewhere far into the future?

**Comment by kroitor** [comment]
> It looks like you assumed that I want to use ccxt keys for user testing, but this is not the case. For user-side testing, I'll make tests with order creation/cancellation logic and that would require some balances in their accounts.

Sounds good!

> You mean that someone sets them somewhere far into the future?

Yes, exactly. Someone is overriding the nonce for microseconds for those keys. We keep multiple lists of keys for testing, most of them non-public.

**Comment by mkutny** [comment]
> Sounds good!

So, just to be sure we're on the same side. Are you comfortable if I extend ccxt integration tests with error handling testing? That would require API keys with trading permissions.

I'm not pushing, if you're not sure I'll be thinking of other ways to achieve that.


> Yes, exactly. Someone is overriding the nonce for microseconds for those keys.

Ahh!!! I remember I submitted few PRs that switched nonce from seconds to milliseconds for some exchanges.

Now imagine those PRs passed Travis but you haven't merged them yet. From now on every commit until merge will fail. So I'm definitely accountable for several of those overrides.

I believe if you switch default nonce from seconds to milliseconds it will ease the problem to some extent.

**Comment by kroitor** [comment]
> Are you comfortable if I extend ccxt integration tests with error handling testing? That would require API keys with trading permissions.

Yes, I am. But some of those keys don't have all permissions (currently). If you find those, let us know, we will re-create the necessary accounts and keys. As long as we test it before merging, we'll be fine.

> Now imagine those PRs passed Travis but you haven't merged them yet. 

I'm merging all of them just in time, so this isn't the case.

> From now on every commit until merge will fail. So I'm definitely accountable for several of those overrides.

Not really, I am talking of creating a new key, then using it for a couple of days, then it gets broken (when another user updates his source and runs his code with nonce overrides). So, it's not your fault.

> I believe if you switch default nonce from seconds to milliseconds it will ease the problem to some extent.

Right, but most noobs don't read our docs before using it. And if we switch the nonce, they will catch a ban right away and will flood the issues-section with questions. Debugging their bans remotely (blindly) would be a real pain, much harder than debugging a nonce issue.

**Comment by mkutny** [comment]
> Yes, I am.

That's absolutely awesome! Thanks! You just saved me a lot of sleepless hours )


> And if we switch the nonce, they will catch a ban right away

That would be the case if we downgraded from milliseconds to seconds. But if we increase the nonce it should not be the problem, should it?

I had to switch few exchanges from seconds to milliseconds because with embedded rate limiter on I occasionally manage to send two requests per second. I believe that those screwers have had the same issue. The only difference is that I submitted PRs and they did not.

**Comment by kroitor** [comment]
> That would be the case if we downgraded from milliseconds to seconds. But if we increase the nonce it should not be the problem, should it?

Well... I would agree, it looks like it would not be a problem at first. But it still turns out to be a problem in the following way: we set nonce to milliseconds(), and they don't have any proper rate-limiting + error handling = they make repeating failing requests in short time in a loop = they catch a ban by running that loop for a few seconds. So the actual reasoning behind using seconds-nonces by default is: we wanted noob-requests to fail in order for them to read and learn, but we don't them to catch a ban and blame us (they will blame us, anyway). 

Turns out, we can't really switch to milliseconds right away (for safety) without fulfilling the rate-limiting, namely these aspects:
- Journaling ([implemented in js](https://github.com/ccxt/ccxt/blob/master/js/base/journal.js), but [not portable yet](https://github.com/ccxt/ccxt/pull/612))
- Repeat-on-recoverable-failure, not implemented in the library yet, we have [some concerns on that](https://github.com/ccxt/ccxt/issues/961), main concern is: this is probably better to leave outside of the lib
- Hard-code [token bucket costs](https://www.kraken.com/en-us/help/api#api-call-rate-limit) for all endpoints:
```
'api': { 
    'private': { 
        'get': [ 
            '/foo', 
            '/bar' 
        ] 
    }
}
```
↓
```
'api': { 
    'private': { 
        'get': { 
            '/foo': { cost: 10, requiresSymbol: true, otherCharacteristics... }, 
            '/bar': 10, 
            ... 
        }
    }
}
```

Let me know if the above makes sense. We are still going to switch to milliseconds eventually anyway (hopefully soon).

> The only difference is that I submitted PRs and they did not.

My personal respect to you for that )

**Comment by mkutny** [comment]
Replaced by #1148
